### PR TITLE
Add indentation to json dump method

### PR DESCRIPTION
**Pull request by sina-rostami**
The problem is specified in issue #747 .

This is a simple solution to improve the dump method without breaking any APIs and is fully backward-compatible.

### CODE REVIEW

File: include/crow/json.h [hunk 1]
```diff
@@ -1826,7 +1826,16 @@ namespace crow
                 out.push_back('"');
             }
 
-            inline void dump_internal(const wvalue& v, std::string& out) const
+            inline void dump_indentation_part(std::string& out, const int indent, const char separator, const int indent_level) const
+            {
+                out.push_back('\n');
+                for (int i = 0; i < indent_level; i++)
```
Comment (gittiver) [inline include/crow/json.h:1832]: out.append(indent *indent_level,separator);

is a simpler replacement

**Review by gittiver — CHANGES_REQUESTED** [review]
Have a look at my comment. As a general comment I tend to leave dump and dump internal as it is and put the indentation into a pretty_print function or something like this.

**Comment by sina-rostami** [comment]
> Have a look at my comment. As a general comment I tend to leave dump and dump internal as it is and put the indentation into a pretty_print function or something like this.

Thank you for the review. Leaving those functions as they are, I should duplicate a bunch of code in `pretty_print` as they have much in common. Moreover, `dump` and `dump_internal` can still be used with and for the `non-pretty` style purpose without losing any performance in comparison to the current version. However, if my reasons didn't convince you, please let me know and I'll separate them.

**Comment by mrozigor** [comment]
I don't know if formatting output should be a part of the library. But that's up too you . Code looks ok I guess ;)

File: include/crow/json.h [hunk 2]
```diff
@@ -1954,13 +1986,20 @@ namespace crow
             }
 
         public:
-            std::string dump() const
+            std::string dump(const int indent, const char separator = ' ') const
```
Comment (mrozigor) [inline include/crow/json.h:1989]: Maybe set a default value also for `indent`?

File: include/crow/json.h [hunk 3]
```diff
@@ -1909,6 +1916,10 @@ namespace crow
                     case type::List:
                     {
                         out.push_back('[');
+
+                        if (indent >= 0)
```
Comment (mrozigor) [inline include/crow/json.h:1920]: Add brackets everywhere.

File: include/crow/json.h (hunk 2 above)
Comment (sina-rostami) [reply include/crow/json.h:1989]: setting a default value for `indent` conflicts with the virtual function `dump` of type `returnable`, so I used to keep the previous interface.

**Review by mrozigor — APPROVED** [review]

**Review by gittiver — APPROVED** [review]

**Pull request merged**